File: android/src/main/kotlin/dev/coolswood/openpanel/OpenpanelPlugin.kt
package dev.coolswood.openpanel

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Entry point of the openpanel plugin.
 *
 * The channel layer is intentionally thin: arguments are unpacked here and
 * handed to [OpenpanelSdk], which owns all state on a single background
 * thread.
 */
class OpenpanelPlugin : FlutterPlugin, MethodCallHandler {
    private var channel: MethodChannel? = null
    private var sdk: OpenpanelSdk? = null
    private var lifecycleTracker: LifecycleTracker? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val openpanelSdk = OpenpanelSdk(binding.applicationContext)
        sdk = openpanelSdk
        lifecycleTracker = LifecycleTracker(binding.applicationContext) { foreground ->
            openpanelSdk.onAppLifecycle(foreground)
        }
        val methodChannel = MethodChannel(binding.binaryMessenger, "openpanel")
        methodChannel.setMethodCallHandler(this)
        channel = methodChannel
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val openpanelSdk = sdk
        if (openpanelSdk == null) {
            result.error("unavailable", "openpanel plugin is not attached", null)
            return
        }

        when (call.method) {
            "initialize" -> {
                val clientId = call.argument<String>("clientId")
                if (clientId.isNullOrEmpty()) {
                    result.error("invalid_arguments", "clientId is required", null)
                    return
                }
                openpanelSdk.initialize(
                    clientId = clientId,
                    clientSecret = call.argument("clientSecret"),
                    apiUrl = call.argument<String>("apiUrl") ?: OpenpanelSdk.DEFAULT_API_URL,
                    automaticTracking = call.argument<Boolean>("automaticTracking") ?: true,
                    disabled = call.argument<Boolean>("disabled") ?: false,
                    verbose = call.argument<Boolean>("verbose") ?: false,
                )
                result.success(null)
            }

            "track" -> {
                openpanelSdk.track(
                    name = call.argument<String>("name") ?: "unknown",
                    properties = call.argument<Map<String, Any?>>("properties"),
                )
                result.success(null)
            }

            "identify" -> {
                val profileId = call.argument<String>("profileId")
                if (profileId.isNullOrEmpty()) {
                    result.error("invalid_arguments", "profileId is required", null)
                    return
                }
                openpanelSdk.identify(
                    profileId = profileId,
                    firstName = call.argument("firstName"),
                    lastName = call.argument("lastName"),
                    email = call.argument("email"),
                    avatar = call.argument("avatar"),
                    properties = call.argument<Map<String, Any?>>("properties"),
                )
                result.success(null)
            }

            "increment", "decrement" -> {
                val profileId = call.argument<String>("profileId")
                val property = call.argument<String>("property")
                if (profileId.isNullOrEmpty() || property.isNullOrEmpty()) {
                    result.error("invalid_arguments", "profileId and property are required", null)
                    return
                }
                val value = call.argument<Number>("value")?.toInt() ?: 1
                openpanelSdk.increment(
                    type = call.method,
                    profileId = profileId,
                    property = property,
                    value = value,
                )
                result.success(null)
            }

            "setGlobalProperties" -> {
                openpanelSdk.setGlobalProperties(call.argument<Map<String, Any?>>("properties"))
                result.success(null)
            }

            "clear" -> {
                openpanelSdk.clear()
                result.success(null)
            }

            "flush" -> {
                openpanelSdk.flush(reason = "manual")
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        lifecycleTracker?.dispose()
        lifecycleTracker = null
        sdk?.dispose()
        sdk = null
    }
}

File: ios/openpanel/Sources/openpanel/OpenpanelPlugin.swift
import Flutter
import UIKit

/// Entry point of the openpanel plugin.
///
/// The channel layer is intentionally thin: arguments are unpacked here and
/// handed to `OpenpanelSdk`, which owns all state on a single serial queue.
public class OpenpanelPlugin: NSObject, FlutterPlugin {
  private let sdk = OpenpanelSdk()
  private var lifecycleTracker: LifecycleTracker?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "openpanel", binaryMessenger: registrar.messenger())
    let instance = OpenpanelPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    instance.start()
  }

  /// Registers lifecycle observers. Kept out of `init` so `sdk` is fully
  /// constructed first.
  private func start() {
    lifecycleTracker = LifecycleTracker { [weak sdk] foreground in
      sdk?.onAppLifecycle(inForeground: foreground)
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]

    switch call.method {
    case "initialize":
      guard let clientId = args["clientId"] as? String, !clientId.isEmpty else {
        result(FlutterError(code: "invalid_arguments", message: "clientId is required", details: nil))
        return
      }
      sdk.initialize(
        clientId: clientId,
        clientSecret: args["clientSecret"] as? String,
        apiUrl: args["apiUrl"] as? String ?? OpenpanelSdk.defaultApiUrl,
        automaticTracking: args["automaticTracking"] as? Bool ?? true,
        disabled: args["disabled"] as? Bool ?? false,
        verbose: args["verbose"] as? Bool ?? false)
      result(nil)

    case "track":
      sdk.track(
        name: args["name"] as? String ?? "unknown",
        properties: args["properties"] as? [String: Any])
      result(nil)

    case "identify":
      guard let profileId = args["profileId"] as? String, !profileId.isEmpty else {
        result(FlutterError(code: "invalid_arguments", message: "profileId is required", details: nil))
        return
      }
      sdk.identify(
        profileId: profileId,
        firstName: args["firstName"] as? String,
        lastName: args["lastName"] as? String,
        email: args["email"] as? String,
        avatar: args["avatar"] as? String,
        properties: args["properties"] as? [String: Any])
      result(nil)

    case "increment", "decrement":
      guard let profileId = args["profileId"] as? String, !profileId.isEmpty,
        let property = args["property"] as? String, !property.isEmpty
      else {
        result(FlutterError(code: "invalid_arguments", message: "profileId and property are required", details: nil))
        return
      }
      sdk.increment(
        type: call.method,
        profileId: profileId,
        property: property,
        value: (args["value"] as? NSNumber)?.intValue ?? 1)
      result(nil)

    case "setGlobalProperties":
      sdk.setGlobalProperties(args["properties"] as? [String: Any] ?? [:])
      result(nil)

    case "clear":
      sdk.clear()
      result(nil)

    case "flush":
      sdk.flush(reason: "manual")
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
